Perf/note store

milesmace opened the pull request:
## Description:
The current implementation uses simple arrays to store notes, which leads to inefficient operations—especially as the dataset grows. Searching for or updating a note in an array requires linear time (`O(n)`), which degrades performance.

This PR replaces arrays with a Map data structure, offering more efficient operations:
- **Search**: `O(1)` average time complexity
- **Insertion**: `O(1)` average time complexity

Using a Map provides faster access and better scalability, improving overall application performance and maintainability.

milesmace committed:
perf: Improve note storage in localStorage

krishnamodepalli commented on client/src/containers/NoteListContainer/NoteList.container.tsx:
```diff
@@ -9,19 +9,20 @@ import { ROUTES } from "@/constants";
 export const NoteListContainer = () => {
   const dispatch = useAppDispatch();
-  const notes = useAppSelector(selectNotes);
+  const noteStore = useAppSelector(selectNotes);
```
update `selectNotes` to `selectNoteStore`

krishnamodepalli commented on client/src/containers/NoteListContainer/NoteList.container.tsx:
```diff
@@ -9,19 +9,20 @@ import { ROUTES } from "@/constants";
       if (!notes) return;
 
-      saveToStorage<INote[]>("notes", notes);
+      saveToStorage<NoteStore>("notes", buildNoteStoreFromArray(notes));
```
Also change the key to store the notes in localStorage to `note-store` or `note_store`. And also we can improve this by using some constants.

krishnamodepalli commented on client/src/features/note/noteSlice.ts:
```diff
@@ -1,32 +1,32 @@
+import { buildNoteStoreFromArray, loadFromStorage } from "@/utils";
 
-const initialState: Note[] = loadFromStorage<Note[]>("notes", []);
+const initialState: NoteStore = loadFromStorage<NoteStore>("notes", {});
```
update everywhere

milesmace committed:
refactor: Add required constants and use them where needed

what's changed:
- Change the theme literals "light", "dark" & localStorage literals
  "notes" into Constants for easier and un-ambiguous code

milesmace committed:
chore: import typo changed

milesmace committed:
Merge branch 'feat/confirm-on-logout' into perf/note-store
